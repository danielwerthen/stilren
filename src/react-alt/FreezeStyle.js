import React from 'react';
import { Consumer } from './context';
import transformElement from './transformElement';

function comp(element, component) {
  function Frozen(props) {
    return React.cloneElement(element, props);
  }
  Frozen.displayName = `Frozen(${component.displayName ||
    component.name ||
    component})`;
  return Frozen;
}

export default class FreezeStyle extends React.Component {
  render() {
    const { children, ...components } = this.props;
    return React.createElement(Consumer, {}, ctx => {
      const result = Object.keys(components).reduce((sum, key) => {
        const item = components[key];
        let component,
          props = {};
        if (Array.isArray(item)) {
          component = item[0];
          props = item[1];
        } else {
          component = item;
        }
        sum[key] = comp(transformElement(component, props, ctx, []), component);
        return sum;
      }, {});

      return children(result);
    });
  }
}
