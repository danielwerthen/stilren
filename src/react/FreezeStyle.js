import React from 'react';
import { Consumer } from './context';
import transformProps from './transformProps';
import { inheritanceStore } from './constants';

function comp(Component, staticProps) {
  function Frozen(props) {
    const finalProps = Object.assign({}, staticProps, props);
    if (props.className && staticProps.className) {
      finalProps.className = `${props.className} ${staticProps.className}`;
    }
    return React.createElement(Component, finalProps);
  }
  Frozen.displayName = `Frozen(${Component.displayName ||
    Component.name ||
    Component})`;
  return Frozen;
}

export default class FreezeStyle extends React.Component {
  render() {
    const { children, ...components } = this.props;
    return React.createElement(Consumer, {}, ctx => {
      const result = Object.keys(components).reduce((sum, key) => {
        const item = components[key];
        let Component,
          props = {};
        if (Array.isArray(item)) {
          Component = item[0];
          props = item[1];
        } else {
          Component = item;
        }
        if (!Component[inheritanceStore]) {
          console.warn('Component is not freezeable', Component);
          return sum;
        }
        const [originalComponent, extensions] = Component[inheritanceStore];
        sum[key] = comp(
          originalComponent,
          transformProps(
            {
              ...(Component.defaultProps || {}),
              ...props,
            },
            ctx,
            extensions
          )
        );
        return sum;
      }, {});

      return children(result);
    });
  }
}
