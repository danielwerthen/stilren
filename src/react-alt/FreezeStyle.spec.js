import React from 'react';
import ReactDOM from 'react-dom/server';
import { Server } from 'styletron-engine-atomic';
import { componentFactory, StilrenProvider } from './index';
import FreezeStyle from './FreezeStyle';

const breakpoints = {
  alpha: '(min-width: 100px)',
  beta: '(max-width: 800px)',
};
const pseudos = ['hover', 'focus'];

function render(Component, props, ...children) {
  const styletron = new Server();
  expect(
    ReactDOM.renderToString(
      React.createElement(
        StilrenProvider,
        { styletron, breakpoints, pseudos },
        React.createElement(Component, props, ...children)
      )
    )
  ).toMatchSnapshot();
  expect(styletron.getCss()).toMatchSnapshot();
}

describe('React createFactory', () => {
  it('should render stuff', () => {
    const MyComponent = componentFactory('div');
    MyComponent.defaultProps = {
      $fontSize: '15px',
    };
    render(
      FreezeStyle,
      {
        Frozen: MyComponent,
      },
      ({ Frozen }) =>
        React.createElement(
          Frozen,
          { $fontSize: '16px', $color: 'red' },
          'Normal props is OK'
        )
    );
  });
});
