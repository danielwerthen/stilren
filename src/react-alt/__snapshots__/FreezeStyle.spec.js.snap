// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React createFactory should render stuff 1`] = `"<div class=\\"ae\\">Normal props is OK</div>"`;

exports[`React createFactory should render stuff 2`] = `".ae{font-size:15px}"`;
